bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieImpl.java

```diff
@@ -935,9 +936,10 @@ private void addEntryInternal(LedgerDescriptor handle, ByteBuf entry,
 
         // journal `addEntry` should happen after the entry is added to ledger storage.
         // otherwise the journal entry can potentially be rolled before the ledger is created in ledger storage.
-        if (masterKeyCache.get(ledgerId) == null) {
+        if (masterKeyCache.getIfPresent(ledgerId) == null) {
             // Force the load into masterKey cache
-            byte[] oldValue = masterKeyCache.putIfAbsent(ledgerId, masterKey);
+            byte[] oldValue = masterKeyCache.getIfPresent(ledgerId);
+            masterKeyCache.put(ledgerId, masterKey);
             if (oldValue == null) {
```

Comment:
If the cache is expired, but ledger is still not rolled, what will happen with below code will executed multiple times
```
getJournal(ledgerId).logAddEntry(bb, false /* ackBeforeSync */, new NopWriteCallback(), null);
```

Reply:
@gaozhangmin it doesn't matter:
the masterKey of the same ledgerID is the same. This is only to prevent multiple writes. If it occurs multiple times, there is no risk. It will only cover the information when the journal does relay.

Reply:
Good point. I have a question. Is there any problem if the `masterKeyCahce` cache is evicted during `readJournal`? I don't see the role of `masterKeyCache` in this process.

Reply:
We should avoid duplicately writing masterKey to journal file,  even duplicate writing has no problem.

Reply:
> Good point. I have a question. Is there any problem if the `masterKeyCahce` cache is evicted during `readJournal`? I don't see the role of `masterKeyCache` in this process.

@wenbingshen 
The main role of masterKeyCache:
1. When addEntry, synchronize a piece of METAENTRY_ID_LEDGER_KEY data, you need masterKeyCache to write only one piece of this data for each ledgerID
2. When the bookie is restarted and the journal is doing data replay, for METAENTRY_ID_LEDGER_KEY, the following operations are done: restore masterKeyCache, restore HandleFactory

Reply:
>We should avoid duplicately writing masterKey to journal file, even it has no problem.
I agree with @gaozhangmin .

Also since this touches the write path I suggest you to make this new cache configurable and disabled by default.

Reply:
if we disable it by default nobody will use it, it will fall into the basket of "experimental features"

Reply:
> if we disable it by default nobody will use it, it will fall into the basket of "experimental features"

have a look @nicoloboschi @gaozhangmin 
I have update the code, when check masterKey exists in cache before write
```
// Force the load into masterKey cache
byte[] oldValue = masterKeyCache.asMap().putIfAbsent(ledgerId, masterKey);
```

Reply:
> if we disable it by default nobody will use it, it will fall into the basket of "experimental features"

@nicoloboschi 
yes,I agree @eolivelli ,  We should use the new logic directly, and the old logic is a bug, which is easy to cause oom